lib/nostrum/struct/permission.ex

```diff
@@ -0,0 +1,182 @@
+          | :use_external_emojis
+  @type voice_permission ::
+          :connect
+          | :speak
+          | :mute_members
+          | :deafen_members
+          | :move_members
+          | :use_vad
+  @type t ::
+          general_permission
+          | text_permission
+          | voice_permission
+
+  @permission_to_bit_map %{
+    create_instant_invite: 0x00000001,
+    kick_members: 0x00000002,
+    ban_members: 0x00000004,
+    administrator: 0x00000008,
+    manage_channels: 0x00000010,
+    manage_guild: 0x00000020,
+    add_reactions: 0x00000040,
+    view_audit_log: 0x00000080,
+    view_channel: 0x00000400,
+    send_messages: 0x00000800,
+    send_tts_messages: 0x00001000,
+    manage_messages: 0x00002000,
+    embed_links: 0x00004000,
+    attach_files: 0x00008000,
+    read_message_history: 0x00010000,
+    mention_everyone: 0x00020000,
+    use_external_emojis: 0x00040000,
+    connect: 0x00100000,
+    speak: 0x00200000,
+    mute_members: 0x00400000,
+    deafen_members: 0x00800000,
+    move_members: 0x01000000,
+    use_vad: 0x02000000,
+    change_nickname: 0x04000000,
+    manage_nicknames: 0x08000000,
+    manage_roles: 0x10000000,
+    manage_webhooks: 0x20000000,
+    manage_emojis: 0x40000000
+  }
+  @bit_to_permission_map Map.new(@permission_to_bit_map, fn {k, v} -> {v, k} end)
+  @permission_list Map.keys(@permission_to_bit_map)
+
+  @doc """
+  Returns `true` if `term` is a permission; otherwise returns `false`.
+
+  ## Examples
+
+  ```Elixir
+  iex> Nostrum.Struct.Permission.is_permission(:administrator)
+  true
+
+  iex> Nostrum.Struct.Permission.is_permission(:not_a_permission)
+  false
+  ```
+  """
+  defguard is_permission(term) when is_atom(term) and term in @permission_list
+
+  @doc """
+  Returns a list of all permissions.
+  """
+  @spec all() :: [t]
+  def all, do: @permission_list
+
+  @doc """
+  Converts the given bit to a permission.
+
+  ## Examples
+
+  ```Elixir
+  iex> Nostrum.Struct.Permission.from_bit(0x04000000)
+  :change_nickname
+  ```
+  """
+  @spec from_bit(bit) :: t
+  def from_bit(bit), do: @bit_to_permission_map[bit]
+
+  @doc """
+  Converts the given bitset to a liist of permissions.
+
+  ## Examples
+
+  ```Elixir
+  iex> Nostrum.Struct.Permission.from_bitset(0x04000000)
+  [:change_nickname]
+
+  iex> Nostrum.Struct.Permission.from_bitset(0x08000002)
+  [:kick_members, :manage_nicknames]
+
+  iex> Nostrum.Struct.Permission.from_bitset(0)
+  []
+  ```
+  """
+  @spec from_bitset(bitset) :: [t]
+  def from_bitset(bitset) do
+    0..53
+    |> Enum.map(fn index -> 0x1 <<< index end)
+    |> Enum.filter(fn mask -> (bitset &&& mask) === mask end)
+    |> Enum.map(fn bit -> from_bit(bit) end)
+  end
+
+  @doc """
+  Converts the given permission to a bit.
+
+  ## Examples
+
+  ```Elixir
+  iex> Nostrum.Struct.Permission.to_bit(:administrator)
+  8
+  ```
+  """
+  @spec to_bit(t) :: bit
+  def to_bit(permission), do: @permission_to_bit_map[permission]
```

Comment:
We have a guard for `is_permission`, don't you think we should be using it here?

Reply:
Addressed in 3943749.